### PR TITLE
Added simple communication with grSim

### LINKED ISSUE #7: Set up simple communication with grSim
Once we have simple skeleton code for each new node in the ROS rewrite, our next step is to set up basic communication with grSim. We would like to be able to send simple commands to grSim that let us control the robots and move them around in the simulation.

**Why?** We want to set up the basics for simulation early, because it will be very useful in verifying and validating future work. It is difficult to write navigation or high-level code without being able to visualize the result in some way.

**Where changes will be made:** All code changes should happen in the `backend_output` node

**Subtasks:**
* Create function to populate data in a grSim Comand protobuf message
* Use boost::asio (a boost networking library) to send these grSim commands over UDP to grSim
* Create a simple wrapper function/interface to set robot velocities in grSim (essentially controlling the robot)

### PR DESCRIPTION
**Pull request by MathewMacDougall**
I added a function that wraps the grSim command protobuf structure to allow us to set a robot's velocity. I encourage everyone to try this out and play around with the temporary values I have set. Keep in mind the values are in the Robot's local coordinates!

I also had some stuttering in grSim where the FPS would drop for a split second. It seems unrelated to these changes but I'm curious if anyone else experiences this.

resolves #7

### CODE REVIEW

File: src/thunderbots/backend_output/grsim/grsim_backend.h [hunk 1]
```diff
@@ -1,13 +1,62 @@
...
+     *            ^
+     *            |
+     *            |
+     *   Y <--- Robot
+     *
+     * @param angular_velocity The angular velocity to set for the robot, in Radians per
+     * second.
+     */
+    void setRobotVelocities(
+        unsigned int robot_id, bool team_colour_yellow, Point velocity,
```
Comment (henryymliu) [inline src/thunderbots/backend_output/grsim/grsim_backend.h]: Is this what clang-format changed this to? Looks a bit funny...

**Review by henryymliu — CHANGES_REQUESTED** [review]

File: src/thunderbots/backend_output/grsim/grsim_backend.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/backend_output/grsim/grsim_backend.h]: Yes. Line length is currently 90. Thoughts on increasing it? I believe CLion's default marker is 120 if that gives some context.

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp [hunk 2]
```diff
@@ -1,11 +1,54 @@
-#include "grsim_backend.h"
+#include "backend_output/grsim/grsim_backend.h"
```
Comment (garethellis0) [inline src/thunderbots/backend_output/grsim/grsim_backend.cpp:1]: Generally, the file name should be the same as the class name (ie. `GrSimBackend.cpp`, `GrSimBackend.h`)

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp [hunk 3]
```diff
@@ -1,11 +1,54 @@
...
+    remote_endpoint = ip::udp::endpoint(ip::address::from_string(network_address), port);
+}
+
+GrSimBackend::~GrSimBackend()
+{
+    socket.close();
+}
+
+void GrSimBackend::setRobotVelocities(
+    unsigned int robot_id, bool team_colour_yellow, Point velocity,
```
Comment (garethellis0) [inline src/thunderbots/backend_output/grsim/grsim_backend.cpp]: We should make an enum for team color:
```
enum TeamColor { YELLOW, BLUE };
```
and then use that here (ie. `bool team_colour_yellow` -> `TeamColor team_color`)

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp [hunk 4]
```diff
@@ -1,11 +1,54 @@
...
+
+    packet.mutable_commands()->set_isteamyellow(team_colour_yellow);
+    packet.mutable_commands()->set_timestamp(0.0);
+    grSim_Robot_Command* command = packet.mutable_commands()->add_robot_commands();
+
+    command->set_id(robot_id);
+
+    // We set a robot velocity, not individual wheel velocities
+    command->set_wheelsspeed(false);
+    command->set_veltangent(static_cast<float>(velocity.x()));
```
Comment (garethellis0) [inline src/thunderbots/backend_output/grsim/grsim_backend.cpp]: Space here would be helpful to make it clear this comment has nothing to do with setting velocity, ie:
```
   // We set a robot velocity, not individual wheel velocities
   command->set_wheelsspeed(false);
   
   command->set_veltangent(static_cast<float>(velocity.x()));
```

File: src/thunderbots/backend_output/grsim/grsim_backend.h (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/backend_output/grsim/grsim_backend.h]: I am strong against increasing the line length further for several reasons
- makes it harder to work with on smaller screens
- using a shorter line length forces people to break things up and makes it more readable

I agree this isn't ideal here, but there are several options in clang-format we can change to make this nicer: `AlignAfterOpenBracket`, `AllowAllParametersOfDeclarationOnNextLine`, `AllowShortFunctionsOnASingleLine`,  `BinPackArguments`, `BinPackParameters`, `IndentWrappedFunctionNames` are some of the ones that can tweak how a function definition is formatted (all are described in more detail here: https://clang.llvm.org/docs/ClangFormatStyleOptions.html)

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp [hunk 5]
```diff
@@ -1,11 +1,54 @@
...
+    socket.send_to(
+        buffer(packet.SerializeAsString(), static_cast<size_t>(packet.ByteSize())),
+        remote_endpoint, 0, err);
 }
 
-void GrSimBackend::sendPrimitives(const std::vector<Primitive> &primitives)
+void GrSimBackend::sendPrimitives(const std::vector<Primitive>& primitives)
 {
-    // TODO: Implement this
+    setRobotVelocities(0, false, Point(0.5, -0.1), Angle::ofRadians(-0.8));
```
Comment (garethellis0) [inline src/thunderbots/backend_output/grsim/grsim_backend.cpp]: This should be left as a `TODO` until we properly implement this.

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp [hunk 6]
```diff
@@ -1,11 +1,54 @@
...
+
+GrSimBackend::~GrSimBackend()
+{
+    socket.close();
+}
+
+void GrSimBackend::setRobotVelocities(
+    unsigned int robot_id, bool team_colour_yellow, Point velocity,
+    Angle angular_velocity)
 {
```
Comment (garethellis0) [inline src/thunderbots/backend_output/grsim/grsim_backend.cpp]: We should make sure we've got unit test coverage for generation of packets to send to GrSim. It's a bit tricky to test network stuff, so we could refactor this to have a 
```
std::string GrSimBackend::getRobotVelocityPacket(
    unsigned int robot_id, bool team_colour_yellow, Point velocity,  Angle angular_velocity)
```
function and test that instead?

File: src/thunderbots/backend_output/grsim/grsim_backend.h [hunk 7]
```diff
@@ -1,13 +1,62 @@
...
    public:
-    explicit GrSimBackend();
+    /**
+     * Creates a new grSim backend
+     *
+     * @param network_address The IP address to publish grSim commands to
+     * @param port The port to publish commands to
+     */
+    explicit GrSimBackend(std::string network_address, unsigned short port);
+    ~GrSimBackend();
```
Comment (garethellis0) [inline src/thunderbots/backend_output/grsim/grsim_backend.h]: nit; but should have empty line above `~GrSimBackend();`

File: src/thunderbots/backend_output/main.cpp [hunk 8]
```diff
@@ -32,8 +33,11 @@ int main(int argc, char** argv)
         node_handle.subscribe("backend/move_prim", 1, movePrimUpdateCallback);
 
     // Initialize variables
-    primitives      = std::vector<Primitive>();
-    Backend backend = GrSimBackend();
+    primitives           = std::vector<Primitive>();
+    GrSimBackend backend = GrSimBackend("127.0.0.1", 20011);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/backend_output/main.cpp]: :+1: for this, I think eventually we should probably have it as a rosparam (basically a runtime-reconfigurable setting)

File: src/thunderbots/backend_output/main.cpp [hunk 9]
```diff
@@ -32,8 +33,11 @@ int main(int argc, char** argv)
...
+
+    // We loop at 30Hz so we don't overload the network with too many packets
+    ros::Rate tick_rate(30);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/backend_output/main.cpp]: similar comment to having the port number as a constant here, eventually would be great to have this as a rosparam

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/thunderbots/backend_output/grsim/grsim_backend.h [hunk 10]
```diff
@@ -1,13 +1,62 @@
...
+
+    /**
+     * Sends the given primitives to grSim
+     * @param primitives the list of primitives to send
+     */
     void sendPrimitives(const std::vector<Primitive> &primitives) override;
```
Comment (garethellis0) [inline src/thunderbots/backend_output/grsim/grsim_backend.h]: If this is defined in `backend.h` we should remove the docstring here. Documentation should only every be defined in a single place.

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/backend_output/grsim/grsim_backend.cpp:1]: Will discuss at meeting today and fix in next PR that deals with interfaces.

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/backend_output/grsim/grsim_backend.cpp]: Will discuss at meeting today and fix in next PR that deals with interfaces

File: src/thunderbots/backend_output/grsim/grsim_backend.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/backend_output/grsim/grsim_backend.h]: We can discuss this in today's meeting. Depending on the rest of the review we can either re-format in this PR or the next one.

File: src/thunderbots/backend_output/grsim/grsim_backend.cpp (hunk 6 above)
Comment (MathewMacDougall) [reply src/thunderbots/backend_output/grsim/grsim_backend.cpp]: I'll separate this function into one that only sends a grSim packet over the network, and one that constructs the packet from parameters. Then we can test the packet construction.

File: src/thunderbots/backend_output/grsim/grsim_backend.h (hunk 1 above)
Comment (henryymliu) [reply src/thunderbots/backend_output/grsim/grsim_backend.h]: Yeah, what I kind of like instead is     
```
void setRobotVelocities(unsigned int robot_id, bool team_colour_yellow, 
                        Point velocity, ...
```

**Review by henryymliu — APPROVED** [review]
LGTM

File: src/thunderbots/backend_output/grsim/grsim_backend.h (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/backend_output/grsim/grsim_backend.h]: @henrymliu - I'm cool with that, didn't sound like anyone had a strong opinion on this, so feel free to change and reformat repo (@MathewMacDougall ?)

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**